### ISSUE TITLE
feature request: remove need for SE_SECURITY_NAME privilege

### ISSUE THREAD

**Issue by bailey27** [dokan-dev/dokany#262]
Hello.

Please forgive me if this issue is a duplicate.  I am fairly new to github.  I searched for my issue and didn't find it.

I have used Dokany to implement the gocryptfs filesystem (https://github.com/rfjakob/gocryptfs) in a Windows application.

My program is called cppcryptfs

https://github.com/bailey27/cppcryptfs

I really wish I didn't need for my app to run as administrator.  It elevates itself by throwing up a UAC dialog, btw.

I found this comment in the mirror.c example

// Add security name privilege. Required here to handle GetFileSecurity
// properly.

All my GetFileSecurity does is encrypt the filenames in the path and prepend the root dir to get at the real path and then it does what mirror.c does which is to open the file and call GetUserObjectSecurity().

Would it be possible to have a way so the application can punt GetFileSecurity back to the driver and let the driver deal with it so the app doesn't need SE_SECURITY_NAME or admin privileges to work?

Maybe the app could return the real path back to the driver and the driver could impersonate the caller and try to get the security information?

I realize that maybe this feature wouldn't help that many apps besides my app and the mirror example avoid needing administrator privileges to work.  But maybe there are more apps that could benefit from this.

I would volunteer to do it myself, but it's been a really long time since I've worked on Windows device drivers.  

Maybe if I have time later, I might have a go at it.

**Comment by Liryna** [comment]
Hello @bailey27 ,

Good job for cppcryptfs !

To disable security request, you can simply set `GetFileSecurity ` and `SetFileSecurity ` to `NULL` here
https://github.com/bailey27/cppcryptfs/blob/master/cppcryptfs/cryptdokan.cpp#L1455-L1456

And remove `AddSeSecurityNamePrivilege` from your code
https://github.com/bailey27/cppcryptfs/blob/master/cppcryptfs/cryptdokan.cpp#L1419-L1426

This will inform the system that you do not have the security features implemented and will ignore it.

**Comment by Rondom** [comment]
I think disabling as Lyrina said is the way to go. 
What you are suggesting sounds impossible or very hard to keep secure (at least as far as I understand). How would you make sure that the functionality you describe is only used for nice use cases an not to work around missing privileges?

**Comment by bailey27** [comment]
Thanks.  It has been a pleasure working with Dokany.

I tried what you said, and works!

I thought it wasn't working, but the problem was running winfstest on a filesystem I created with my program running the old way (as admin, acquiring se name, and exporting Get/SeFileSecurity).

When I run it without admin and without exporting Get/SetFileSecurity, and I make a new filesystem, it passes the 169/171 tests like it did before.

So now I  made it so it tries to get se name, and if not, it sets those functions to NULL.

Thank you very much!

**Comment by Liryna** [comment]
:+1: Thank you for your feedback !
When you FS is ready, tell me and I will add it on our web page !

**Comment by bailey27** [comment]
It doesn't really work.

When I didn't run as administrator and set the Get & SetFileSecurity callbacks to NULL, I couldn't copy files OUT of my filesystem.  But I could copy them in.

I couldn't copy them out with either the copy command in a cmd shell, or with file explorer.

I could copy them out with the cygwin cp command.

I could copy (with copy cmd) a file to an existing file within my fs.  But I couldn't copy a file to a new file within my fs.

It looks like whatever cmd and explorer does (CopyFile()?) has to be able to Get (at least some of) the file security.

I looked at the docs for GetFileSecurity(), and it looks like it doesn't always need SE_SECURITY_NAME depending on what it's asked for.

I tried running not as administrator but exporting the Get/SetFileSecurity() callbacks, and my fs is working better. The file copy seems to be ok again.

I'll keep testing.
